cloudstate/cloudstate.go

```diff
@@ -0,0 +1,249 @@
+		ServiceInfo: &protocol.ServiceInfo{
+			ServiceName:           options.ServiceName,
+			ServiceVersion:        options.ServiceVersion,
+			ServiceRuntime:        fmt.Sprintf("%s %s/%s", runtime.Version(), runtime.GOOS, runtime.GOARCH),
+			SupportLibraryName:    SupportLibraryName,
+			SupportLibraryVersion: SupportLibraryVersion,
+		},
+	}
+	responder.fileDescriptorSet = &filedescr.FileDescriptorSet{
+		File: make([]*filedescr.FileDescriptorProto, 0),
+	}
+	return responder
+}
+
+// Discover returns an entity spec for
+func (r *EntityDiscoveryResponder) Discover(c context.Context, pi *protocol.ProxyInfo) (*protocol.EntitySpec, error) {
+	log.Printf("Received discovery call from sidecar [%s w%s] supporting CloudState %v.%v\n",
+		pi.ProxyName,
+		pi.ProxyVersion,
+		pi.ProtocolMajorVersion,
+		pi.ProtocolMinorVersion,
+	)
+	for _, filename := range []string{
+		"google/protobuf/empty.proto",
+		"google/protobuf/any.proto",
+		"google/protobuf/descriptor.proto",
+		"google/api/annotations.proto",
+		"google/api/http.proto",
+		"cloudstate/event_sourced.proto",
+		"cloudstate/entity.proto",
+		"cloudstate/entity_key.proto",
+	} {
+		if err := r.registerFileDescriptorProto(filename); err != nil {
+			return nil, err
+		}
+	}
+	log.Printf("Responding with: %v\n", r.entitySpec.GetServiceInfo())
+	return r.entitySpec, nil
+}
+
+// ReportError logs any user function error reported by the CloudState proxy.
+func (r *EntityDiscoveryResponder) ReportError(c context.Context, fe *protocol.UserFunctionError) (*empty.Empty, error) {
+	log.Printf("ReportError: %v\n", fe)
+	return &empty.Empty{}, nil
+}
+
+func (r *EntityDiscoveryResponder) updateSpec() (err error) {
+	protoBytes, err := proto.Marshal(r.fileDescriptorSet)
+	if err != nil {
+		return errors.New("unable to Marshal FileDescriptorSet")
+	}
+	r.entitySpec.Proto = protoBytes
+	return nil
+}
+
+func (r *EntityDiscoveryResponder) resolveFileDescriptors(dc DescriptorConfig) error {
+	// service
+	if dc.Service != "" {
+		if err := r.registerFileDescriptorProto(dc.Service); err != nil {
+			return err
+		}
+	} else {
+		if dc.ServiceMsg != nil {
+			if err := r.registerFileDescriptor(dc.ServiceMsg); err != nil {
+				return err
+			}
+		}
+	}
+	// and dependent domain descriptors
+	for _, dp := range dc.Domain {
+		if err := r.registerFileDescriptorProto(dp); err != nil {
+			return err
+		}
+	}
+	for _, dm := range dc.DomainMessages {
+		if err := r.registerFileDescriptor(dm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *EntityDiscoveryResponder) registerEntity(e *EventSourcedEntity, config DescriptorConfig) error {
+	if err := r.resolveFileDescriptors(config); err != nil {
+		return fmt.Errorf("failed to resolveFileDescriptor for DescriptorConfig: %+v: %w", config, err)
+	}
+	persistenceID := e.entityName
+	if e.PersistenceID != "" {
+		persistenceID = e.PersistenceID
+	}
+	r.entitySpec.Entities = append(r.entitySpec.Entities, &protocol.Entity{
+		EntityType:    EventSourced,
```

Comment:
I think we can't. Like for the other comments about a type safe way to get the service descriptor, we can't (see below the example for _EventSourced_serviceDesc). I explained this also in the then Draft - PR: https://github.com/cloudstateio/cloudstate/pull/103#issuecomment-531747337 on how to access this information. The comment also mentions a few partly open Issues on the gRPC project for Go.

event_sourced.pb.go
```
var _EventSourced_serviceDesc = grpc.ServiceDesc{
	ServiceName: "cloudstate.eventsourced.EventSourced",
	HandlerType: (*EventSourcedServer)(nil),
	Methods:     []grpc.MethodDesc{},
	Streams: []grpc.StreamDesc{
		{
			StreamName:    "handle",
			Handler:       _EventSourced_Handle_Handler,
			ServerStreams: true,
			ClientStreams: true,
		},
	},
	Metadata: "cloudstate/event_sourced.proto",
}
```

@jsravn do you see any other way to do that? Tho, I've seen the https://github.com/gogo/protobuf project and that it is more open in this regard, but I have not tried it.

Reply:
I think gogo is only protobuf 2? It's been a while since I looked at it.

It does seem like golang protobuf is fairly anemic, unfortunately.

But I think we can do this with protoreflect:

```
func (r *EntityDiscoveryService) registerEntity(e *EventSourcedEntity, config DescriptorConfig) error {
	if err := r.resolveFileDescriptors(config); err != nil {
		return fmt.Errorf("failed to resolveFileDescriptor for DescriptorConfig: %+v: %w", config, err)
	}
	persistenceID := e.entityName
	if e.PersistenceID != "" {
		persistenceID = e.PersistenceID
	}
	fd, err := desc.LoadFileDescriptor("cloudstate/event_sourced.proto")
	if err != nil {
		return err
	}
	r.entitySpec.Entities = append(r.entitySpec.Entities, &protocol.Entity{
		EntityType:    fd.FindService("EventSourced").GetFullyQualifiedName(),
		ServiceName:   e.ServiceName,
		PersistenceId: persistenceID,
	})
	return r.updateSpec()
}
```